panther_docking/src/panther_charging_dock.cpp

```diff
@@ -0,0 +1,420 @@
+  node_->get_parameter(name + ".staging_x_offset", staging_x_offset_);
+  node_->get_parameter(name + ".staging_yaw_offset", staging_yaw_offset_);
+
+  node_->get_parameter(
+    name + ".enable_charger_service_call_timeout", enable_charger_service_call_timeout_);
+
+  double filter_coef;
+  node_->get_parameter(name + ".filter_coef", filter_coef);
+  filter_ = std::make_unique<opennav_docking::PoseFilter>(filter_coef, external_detection_timeout_);
+}
+
+void PantherChargingDock::cleanup()
+{
+  dock_pose_pub_.reset();
+  staging_pose_pub_.reset();
+
+  if (panther_utils::common_utilities::IsPantherVersionAtLeast(panther_version_, 1.2f)) {
+    charging_status_sub_.reset();
+    io_state_sub_.reset();
+    charger_enable_client_.reset();
+  }
+}
+
+void PantherChargingDock::activate()
+
+{
+  dock_pose_pub_ = node_->create_publisher<geometry_msgs::msg::PoseStamped>("docking/dock_pose", 1);
+  staging_pose_pub_ = node_->create_publisher<geometry_msgs::msg::PoseStamped>(
+    "docking/staging_pose", 1);
+
+  if (panther_utils::common_utilities::IsPantherVersionAtLeast(panther_version_, 1.2f)) {
+    charging_status_sub_ = node_->create_subscription<panther_msgs::msg::ChargingStatus>(
+      "battery/charging_status", rclcpp::SystemDefaultsQoS(),
+      [this](const std::shared_ptr<panther_msgs::msg::ChargingStatus> msg) -> void {
+        charging_status_box_.set(std::move(msg));
+      });
+    io_state_sub_ = node_->create_subscription<panther_msgs::msg::IOState>(
+      "hardware/io_state", rclcpp::SystemDefaultsQoS(),
+      [this](const std::shared_ptr<panther_msgs::msg::IOState> msg) -> void {
+        io_state_box_.set(std::move(msg));
+      });
+
+    charger_enable_client_ =
+      node_->create_client<std_srvs::srv::SetBool>("hardware/charger_enable");
+
+    using namespace std::chrono_literals;
+    if (!charger_enable_client_->wait_for_service(1s)) {
+      RCLCPP_WARN_STREAM(
+        logger_,
+        "Service \"hardware/charger_enable\" not available. Make sure if Panther ROS 2 Stack is "
+        "running.");
+    }
+  }
+}
+
+void PantherChargingDock::deactivate()
+{
+  dock_pose_pub_.reset();
+  staging_pose_pub_.reset();
+
+  if (panther_utils::common_utilities::IsPantherVersionAtLeast(panther_version_, 1.2f)) {
+    charging_status_sub_.reset();
+    io_state_sub_.reset();
+    charger_enable_client_.reset();
+  }
+}
+
+geometry_msgs::msg::PoseStamped PantherChargingDock::getStagingPose(
+  const geometry_msgs::msg::Pose & pose, const std::string & frame)
+{
+  // When the pose if default the robot is docking so the frame is the dock frame
+  if (pose == geometry_msgs::msg::Pose()) {
+    dock_frame_ = frame;
+    updateDockPoseAndPublish();
+    updateStagingPoseAndPublish(base_frame_name_);
+  } else {
+    if (dock_frame_.empty()) {
+      throw opennav_docking_core::FailedToControl("Cannot undock before docking!");
+    }
+    updateDockPoseAndPublish();
+    updateStagingPoseAndPublish(frame);
+  }
```

Comment:
Why not to invoke 
```cpp
    updateDockPoseAndPublish();
    updateStagingPoseAndPublish(frame);
```
once after the if-else statement and only define the frame to be used in  `updateStagingPoseAndPublish()`?